src/rust/include/tracing.h

```diff
@@ -0,0 +1,252 @@
+#ifdef WIN32
+typedef uint16_t codeunit;
+#else
+typedef uint8_t codeunit;
+#endif
+
+struct TracingHandle;
+typedef struct TracingHandle TracingHandle;
+
+/// Initializes the tracing crate, returning a handle for the logging
+/// component. The handle must be freed to close the logging component.
+///
+/// If log_path is NULL, logging is sent to standard output.
+TracingHandle* tracing_init(
+    const codeunit* log_path,
+    size_t log_path_len,
+    const char* initial_filter,
+    bool log_timestamps);
+
+/// Frees a tracing handle returned from `tracing_init`;
+void tracing_free(TracingHandle* handle);
+
+/// Reloads the tracing filter.
+void tracing_reload(TracingHandle* handle, const char* new_filter);
+
+struct TracingCallsite;
+typedef struct TracingCallsite TracingCallsite;
+
+/// Creates a tracing callsite.
+///
+/// You should usually call the `TracingLog` macro (or one of the helper
+/// macros such as `TracingInfo`) instead of calling this directly.
+///
+/// This MUST ONLY be called to assign a `static void*`, and all string
+/// arguments MUST be `static const char*`.
+TracingCallsite* tracing_callsite(
+    const char* name,
+    const char* target,
+    const char* level,
+    const char* file,
+    uint32_t line,
+    const char** fields,
+    size_t fields_len,
+    bool is_span);
+
+struct TracingSpanHandle;
+typedef struct TracingSpanHandle TracingSpanHandle;
+
+struct TracingSpanGuard;
+typedef struct TracingSpanGuard TracingSpanGuard;
+
+/// Creates a span for a callsite.
+///
+/// The span must be freed when it goes out of scope.
+TracingSpanHandle* tracing_span_create(const TracingCallsite* callsite);
+
+/// Clones the given span.
+///
+/// Both spans need to be separately freed when they go out of scope.
+TracingSpanHandle* tracing_span_clone(const TracingSpanHandle* span);
+
+/// Frees a span.
+void tracing_span_free(TracingSpanHandle* span);
+
+/// Enters the given span, returning a guard.
+///
+/// `tracing_span_exit` must be called to drop this guard before the span
+/// goes out of scope.
+TracingSpanGuard* tracing_span_enter(const TracingSpanHandle* span);
+
+/// Exits a span by dropping the given guard.
+void tracing_span_exit(TracingSpanGuard* guard);
+
+/// Logs a message for a callsite.
+///
+/// You should usually call the `TracingLog` macro (or one of the helper
+/// macros such as `TracingInfo`) instead of calling this directly.
+void tracing_log(
+    const TracingCallsite* callsite,
+    const char** field_values,
+    size_t fields_len);
+
+#ifdef __cplusplus
+}
+#endif
+
+//
+// Helper macros
+//
+
+#define T_DOUBLEESCAPE(a) #a
+#define T_ESCAPEQUOTE(a) T_DOUBLEESCAPE(a)
+
+// Computes the length of the given array. This is COUNT_OF from Chromium.
+#define T_ARRLEN(x) ((sizeof(x) / sizeof(0 [x])) / ((size_t)(!(sizeof(x) % sizeof(0 [x])))))
+
+// Constructs a tracing callsite.
+//
+// name, target, and fields MUST be static constants, and the output of this
+// macro MUST be stored as a static void*.
```

Comment:
```suggestion
// macro MUST be stored as a static TracingCallsite*.
```

Reply:
(This change is included in my diff for the compile-time checks.)